README.md
<!-- badges: start -->
[![Lifecycle: Experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![Project Status: WIP - Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)
[![Code size](https://img.shields.io/github/languages/code-size/sfcheung/out2text.svg)](https://github.com/sfcheung/out2text)
[![Last Commit at Main](https://img.shields.io/github/last-commit/sfcheung/out2text.svg)](https://github.com/sfcheung/out2text/commits/main)
[![R-CMD-check](https://github.com/sfcheung/out2text/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/sfcheung/out2text/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

(Version 0.0.0.9002, updated on 2023-03-12, [release history](https://sfcheung.github.io/out2text/news/index.html))

# out2text <img src="man/figures/logo.png" align="right" height="150" />

(Work-in-progress. Not ready for use)

This package contain functions to extract values such as
regression coefficients, *p*-values, confidence intervals,
from common procedures such as regression conducted by `lm()`
and format them as strings so that they can be embedded in
RMarkdown or Quarto documents.

For more information on this package, please visit its GitHub page:

https://sfcheung.github.io/out2text/

# Installation

Stable release versions of this package can be downloaded below:

https://github.com/sfcheung/out2text/releases

The latest developmental version of this package can be installed by `remotes::install_github`:

```r
remotes::install_github("sfcheung/out2text")
```

# Issues

If you have any suggestions and found any bugs, please feel
feel to open a GitHub issue. Thanks.

NEWS.md
# out2text 0.0.0.9002

* Added a `NEWS.md` file to track changes to the package. (0.0.0.9000)
* Added some methods for `lavaan`-class objects. (0.0.0.9000)
* Fixed the pkgdown site. (0.0.0.9001)
* Added `text_rsq_p()` and `text_rsq_f()` for F statistic
  and its *p*-value in `lm()`. (0.0.0.9002)

tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(out2text)

test_check("out2text")


tests/testthat/test_rsq.R
library(testthat)
library(out2text)

set.seed(894252)
n <- 10
x <- rnorm(n)
y <- .6 * x + rnorm(n, 0, .8)

lm_out <- lm(y ~ x)

text_rsq(lm_out)
test_that("lm", {
    expect_identical(text_rsq(lm_out),
                     "0.624")
  })

glm_out <- glm(y ~ x)

test_that("glm", {
    expect_error(text_rsq(glm_out))
  })

test_that("NULL", {
    expect_identical(text_rsq(n), NULL)
  })


test_that("F statistics", {
    expect_identical(text_rsq_f(lm_out),
                     "F(1, 8) = 13.254")
  })

test_that("F statistics", {
    expect_identical(text_rsq_p(lm_out),
                     "p = 0.007")
  })


tests/testthat/test_coef.R
library(testthat)
library(out2text)

set.seed(894252)
n <- 10
x <- rnorm(n)
y <- .6 * x + rnorm(n, 0, .8)

lm_out <- lm(y ~ x)

glm_out <- glm(y ~ x)

test_that("lm and glm", {
    expect_identical(text_coef(lm_out, param = "x"),
                     "0.433")
    expect_identical(text_coef(glm_out, param = "x"),
                     "0.433")
  })


library(lavaan)
mod <-
"
visual  =~ x1 + x2 + x3
textual =~ x4 + x5 + x6
speed   =~ x7 + x8 + x9
textual ~ c(a1, a2) * visual
speed ~ c(b1, b2) * textual
"
fit <- sem(mod, data = HolzingerSwineford1939, group = "school")
est <- parameterEstimates(fit)
std <- standardizedSolution(fit)

test_that("lavaan", {
    expect_identical(text_coef(fit, param = "visual=~x3"),
                     "0.543")
    expect_identical(text_coef(fit, param = "textual~visual", standardized = TRUE),
                     "0.485")
    expect_identical(text_coef(fit, param = "visual=~x3", group = 2, standardized = TRUE),
                     "0.729")
    expect_identical(text_coef(fit, param = "x6 ~1", group = 2),
                     "2.469")
  })

# text_coef(n)


R/text_rsq.R
#' @title A R-square as Text
#'
#' @description Extracts the R-square from an object
#' and returns it as a text string.
#'
#' @details Details
#'   (Include subjects for verbs.)
#'   (Use 3rd person forms for verbs.)
#'
#' @return A one-element character vector.
#'
#' @param x A supported object with an R-square in the results.
#'
#' @param digits The number of decimal places to keep. To be
#' passed to [formatC()].
#'
#' @param format The format string used in [formatC()]. Default
#' is `"f"`.
#'
#' @param type The type of R-squared to be extracted.
#' Either `"unadjusted"`, the default, or `"adjusted"`,
#' which returns the adjusted R-squared.
#'
#'
#' @param ...  Optional arguments. Ignored for now.
#'
#'
#' @author Shu Fai Cheung <https://orcid.org/0000-0002-9871-9448>
#'
#'
#' @examples
#'
#' set.seed(894252)
#' n <- 10
#' x <- rnorm(n)
#' y <- .6 * x + rnorm(n, 0, .8)
#' out <- lm(y ~ x)
#' text_rsq(out)
#'
#' @export

text_rsq <- function(x, ...) UseMethod("text_rsq")

#' @export
#' @rdname text_rsq

text_rsq.default <- function(x,
                             digits = getOption("out2text_coef_digits", 3),
                             format = getOption("out2text_coef_format", "f"),
                             ...) {
    NULL
  }


#' @export
#' @rdname text_rsq

text_rsq.lm <- function(x,
                        digits = getOption("out2text_coef_digits", 3),
                        format = getOption("out2text_coef_format", "f"),
                        type = c("unadjusted", "adjusted"),
                        ...) {
    type <- match.arg(type)
    out <- switch(type,
                  unadjusted = summary(x)$r.squared,
                  adjusted = summary(x)$adj.r.squared)
    if (is.null(out)) {
        stop("The summary of the object does not have the requested R-squared.")
      }
    out <- formatC(out,
                   digits = digits,
                   format = format,
                   ...)
    names(out) <- NULL
    out
  }


R/text_coef.R
#' @title A Coefficient as Text
#'
#' @description Extracts a coefficient from an object
#' and returns it as a text string.
#'
#' @details Details
#'   (Include subjects for verbs.)
#'   (Use 3rd person forms for verbs.)
#'
#' @return A one-element character vector.
#'
#' @param x A supported object with coefficients in the results.
#' @param ...  Optional arguments. Ignored for now.
#'
#'
#' @author Shu Fai Cheung <https://orcid.org/0000-0002-9871-9448>
#'
#'
#' @examples
#'
#' set.seed(894252)
#' n <- 10
#' x <- rnorm(n)
#' y <- .6 * x + rnorm(n, 0, .8)
#' out <- lm(y ~ x)
#' text_coef(out, "x")
#'
#' @export

text_coef <- function(x, ...) UseMethod("text_coef")

#' @param param The name of the term or parameter for which
#' the coefficients will be returned.
#' @param digits The number of decimal places to keep. To be
#' passed to [formatC()].
#' @param format The format string used in [formatC()]. Default
#' is `"f"`.
#' @export
#' @rdname text_coef

text_coef.default <- function(x,
                          param = NULL,
                          digits = getOption("out2text_coef_digits", 3),
                          format = getOption("out2text_coef_format", "f"),
                          ...) {
    if (is.null(param) || length(param) != 1 ||
        !is.character(param)) {
        stop("param must be a length 1 character.")
      }
    coef0 <- tryCatch(stats::coef(x),
                    error = function(e) e)
    if (inherits(coef0, "error")) {
        stop("coef() failed to extract coefficients.")
      }
    coef1 <- coef0[param]
    out <- formatC(coef1,
                    digits = digits,
                    format = format,
                    ...)
    names(out) <- NULL
    out
  }


#' @export
#' @rdname text_coef

text_coef.lm <- function(x,
                          param = NULL,
                          digits = getOption("out2text_coef_digits", 3),
                          format = getOption("out2text_coef_format", "f"),
                          ...) {
    NextMethod()
  }

#' @export
#' @rdname text_coef

text_coef.glm <- function(x,
                          param = NULL,
                          digits = getOption("out2text_coef_digits", 3),
                          format = getOption("out2text_coef_format", "f"),
                          ...) {
    NextMethod()
  }

#' @param standardized Logical. Whether the estimate in the standardized
#' solution is to be returned. Default is `FALSE`.
#' @param group For an `lavaan`-class object, the group from which the
#' estimate is to be returned. Default is 1.
#'
#' @export
#' @rdname text_coef

text_coef.lavaan <- function(x,
                             param = NULL,
                             digits = getOption("out2text_coef_digits", 3),
                             format = getOption("out2text_coef_format", "f"),
                             standardized = FALSE,
                             group = 1L,
                             ...) {
    param_str <- lavaan::lavParseModelString(
                    model.syntax = param,
                    as.data.frame. = TRUE,
                    warn = FALSE)
    if (standardized) {
        ct <- lavaan::standardizedSolution(x,
                                           se = FALSE,
                                           zstat = FALSE,
                                           pvalue = FALSE,
                                           ci = FALSE,
                                           ...)
        ct$est <- ct$est.std
      } else {
        ct <- lavaan::parameterEstimates(x, ...)
      }
    if (lavaan::lavInspect(x, "ngroups") == 1) {
        ct$group <- 1
      }
    coef1 <- ct[(ct$lhs == param_str$lhs) &
                (ct$op == param_str$op) &
                (ct$rhs == param_str$rhs) &
                (ct$group == group), "est"]
    out <- formatC(coef1,
                   digits = digits,
                   format = format,
                   ...)
    names(out) <- NULL
    out
  }

R/text_rsq_f.R
#' @title The F Statistics of R-square as Text
#'
#' @description Extracts the F statistics of
#' the R-square from an object
#' and returns it as a text string.
#'
#' @details Details
#'   (Include subjects for verbs.)
#'   (Use 3rd person forms for verbs.)
#'
#' @return A one-element character vector.
#'
#' @param x A supported object with a F statistic the results.
#'
#' @param digits The number of decimal places to keep. To be
#' passed to [formatC()].
#'
#' @param format The format string used in [formatC()]. Default
#' is `"f"`.
#'
#' @param italics Logical. Whether the statistics should be
#' italicized (using markdown). Default is `FALSE`.
#'
#' @param ...  Optional arguments. Ignored for now.
#'
#'
#' @author Shu Fai Cheung <https://orcid.org/0000-0002-9871-9448>
#'
#'
#' @examples
#'
#' set.seed(894252)
#' n <- 10
#' x <- rnorm(n)
#' y <- .6 * x + rnorm(n, 0, .8)
#' out <- lm(y ~ x)
#' text_rsq_f(out)
#'
#' @export

text_rsq_f <- function(x, ...) UseMethod("text_rsq_f")

#' @export
#' @rdname text_rsq_f

text_rsq_f.default <- function(x,
                             digits = getOption("out2text_coef_digits", 3),
                             format = getOption("out2text_coef_format", "f"),
                             italics = FALSE,
                             ...) {
    NULL
  }


#' @export
#' @rdname text_rsq_f

text_rsq_f.lm <- function(x,
                        digits = getOption("out2text_coef_digits", 3),
                        format = getOption("out2text_coef_format", "f"),
                        italics = FALSE,
                        ...) {
    fstat <- summary(x)$fstatistic
    if (is.null(fstat)) {
        stop("The summary of the object does not have the F statistic.")
      }
    out <- text_f(fstat,
                  digits = digits,
                  format = format,
                  italics = italics,
                   ...)
    names(out) <- NULL
    out
  }

#' @noRd

text_f <- function(fstat,
                   df1,
                   df2,
                   digits = getOption("out2text_test_digits", 3),
                   format = getOption("out2text_test_format", "f"),
                   italics = FALSE,
                   ...) {
    if (length(fstat == 3)) {
        fstat_old <- fstat
        df2 <- fstat[3]
        df1 <- fstat[2]
        fstat <- fstat[1]
      }
    out <- paste0(ifelse(italics, "*F*", "F"),
                  "(",
                  df1,
                  ", ",
                  df2,
                  ")",
                  " = ",
                  formatC(fstat,
                          digits = digits,
                          format = format,
                          ...)
                  )
    out
  }

R/text_coef_p.R
#' @title The P-Value of a Coefficient as Text
#'
#' @description Extracts the p-value of a coefficient from
#' an object
#' and returns it as a text string.
#'
#' @details Details
#'   (Include subjects for verbs.)
#'   (Use 3rd person forms for verbs.)
#'
#' @return A one-element character vector.
#'
#' @param x A supported object with coefficients in the
#' results.
#' @param param The name of the term or parameter for
#' which
#' the coefficients will be returned.
#' @param digits The number of decimal places to keep.
#' To be
#' passed to [formatC()].
#' @param format The format string used in [formatC()].
#' Default
#' is `"f"`.
#' @param ...  Optional arguments. Ignored for now.
#'
#'
#' @author Shu Fai Cheung <https://orcid.org/0000-0002-9871-9448>
#'
#'
#' @examples
#'
#' set.seed(894252)
#' n <- 10
#' x <- rnorm(n)
#' y <- .6 * x + rnorm(n, 0, .8)
#' out <- lm(y ~ x)
#' text_coef_p(out, "x")
#'
#' @export

text_coef_p <- function(x, ...) UseMethod("text_coef_p")

#' @export
#' @rdname text_coef_p

text_coef_p.default <- function(x,
                                param = NULL,
                                digits = getOption("out2text_coef_p_digits", 3),
                                format = getOption("out2text_coef_p_format", "f"),
                                ...) {
    NULL
  }


#' @export
#' @rdname text_coef_p

text_coef_p.lm <- function(x,
                           param = NULL,
                           digits = getOption("out2text_coef_p_digits", 3),
                           format = getOption("out2text_coef_p_format", "f"),
                          ...) {
    if (is.null(param) || length(param) != 1 ||
        !is.character(param)) {
        stop("param must be a length 1 character.")
      }
    coefm <- summary(x)$coefficients
    out0 <- coefm[param, "Pr(>|t|)"]
    out <- format_pval(out0, digits = digits,
                       format = format)
    out
  }

#' @export
#' @rdname text_coef_p

text_coef_p.glm <- function(x,
                            param = NULL,
                            digits = getOption("out2text_coef_p_digits", 3),
                            format = getOption("out2text_coef_p_format", "f"),
                            ...) {
    NextMethod()
  }

#' @param standardized Logical. Whether the estimate in the standardized
#' solution is to be returned. Default is `FALSE`.
#' @param group For an `lavaan`-class object, the group from which the
#' estimate is to be returned. Default is 1.
#' @export
#' @rdname text_coef_p

text_coef_p.lavaan <- function(x,
                               param = NULL,
                               digits = getOption("out2text_coef_p_digits", 3),
                               format = getOption("out2text_coef_p_format", "f"),
                               standardized = FALSE,
                               group = 1L,
                               ...) {
    param_str <- lavaan::lavParseModelString(
                    model.syntax = param,
                    as.data.frame. = TRUE,
                    warn = FALSE)
    if (standardized) {
        ct <- lavaan::standardizedSolution(x,
                                           se = TRUE,
                                           zstat = TRUE,
                                           pvalue = TRUE,
                                           ci = FALSE,
                                           ...)
        ct$est <- ct$est.std
      } else {
        ct <- lavaan::parameterEstimates(x, ...)
      }
    if (lavaan::lavInspect(x, "ngroups") == 1) {
        ct$group <- 1
      }
    coef1 <- ct[(ct$lhs == param_str$lhs) &
                (ct$op == param_str$op) &
                (ct$rhs == param_str$rhs) &
                (ct$group == group), "pvalue"]
    out <- format_pval(coef1, digits = digits,
                       format = format)
    out
  }

#' @noRd

format_pval <- function(x,
                        digits = 3,
                        format = "f") {
    pmin <- 10^(-digits)
    if (x < pmin) {
        out <- paste0("p < ",
                      formatC(pmin, digits = digits,
                              format = format))
      } else {
        out <- paste0("p = ",
                      formatC(x, digits = digits,
                              format = format))
      }
    return(out)
  }

R/text_confint.R
#' @title A Confidence Interval as Text
#'
#' @description Extracts a confidence interval from an object
#' and returns the interval as a text string.
#'
#' @details Details
#'   (Include subjects for verbs.)
#'   (Use 3rd person forms for verbs.)
#'
#' @return A one-element character vector of the interval.
#'
#' @param x A supported object with coefficients in the results.
#' @param ...  Optional arguments. To be passed to
#' [stats::confint()].
#'
#'
#' @author Shu Fai Cheung <https://orcid.org/0000-0002-9871-9448>
#'
#'
#' @examples
#'
#' set.seed(894252)
#' n <- 10
#' x <- rnorm(n)
#' y <- .6 * x + rnorm(n, 0, .8)
#' out <- lm(y ~ x)
#' text_confint(out, "x")
#'
#' @export

text_confint <- function(x, ...) UseMethod("text_confint")

#' @param param The name of the term or parameter for which
#' the coefficients will be returned.
#' @param level The level of confidence. Default is .95.
#' @param digits The number of decimal places to keep. To be
#' passed to [formatC()].
#' @param format The format string used in [formatC()]. Default
#' is `"f"`.
#' @param sep The separator. Default is `", "`.
#' @param brackets The brackets. A character vector with
#' two elements. Default is `c("[", "]")`.
#' @export
#' @rdname text_confint

text_confint.default <- function(x,
                          param = NULL,
                          level = .95,
                          digits = getOption("out2text_coef_digits", 3),
                          format = getOption("out2text_coef_format", "f"),
                          sep = getOption("out2text_confint_sep", ", "),
                          brackets = getOption("out2text_confint_brackets", c("[", "]")),
                          ...) {
    if (is.null(param) || length(param) != 1 ||
        !is.character(param)) {
        stop("param must be a length 1 character.")
      }
    ci0 <- tryCatch(suppressMessages(stats::confint(object = x,
                                   parm = param,
                                   level = level,
                                   ...)),
                    error = function(e) e)
    if (inherits(ci0, "error")) {
        stop("confint() failed to extract coefficients.")
      }
    if (is.null(dim(ci0))) {
        out0 <- ci0
      } else {
        if (nrow(ci0) == 1) {
            out0 <- ci0[1, , drop = TRUE]
          } else {
            out0 <- ci0[param, , drop = TRUE]
          }
      }
    out1 <- formatC(out0,
                    digits = digits,
                    format = format)
    out <- paste0(brackets[1],
                  paste0(out1, collapse = sep),
                  brackets[2])
    out
  }


#' @export
#' @rdname text_confint

text_confint.lm <- function(x,
                          param = NULL,
                          level = .95,
                          digits = getOption("out2text_coef_digits", 3),
                          format = getOption("out2text_coef_format", "f"),
                          sep = getOption("out2text_confint_sep", ", "),
                          brackets = getOption("out2text_confint_brackets", c("[", "]")),
                          ...) {
    NextMethod()
  }

#' @export

text_confint.glm <- function(x,
                      param = NULL,
                      level = .95,
                      digits = getOption("out2text_coef_digits", 3),
                      format = getOption("out2text_coef_format", "f"),
                      sep = getOption("out2text_confint_sep", ", "),
                      brackets = getOption("out2text_confint_brackets", c("[", "]")),
                      ...) {
    NextMethod()
  }

#' @param standardized Logical. Whether the estimate in the standardized
#' solution is to be returned. Default is `FALSE`.
#' @param group For an `lavaan`-class object, the group from which the
#' estimate is to be returned. Default is 1.
#' @export
#' @rdname text_confint

text_confint.lavaan <- function(x,
                        param = NULL,
                        level = .95,
                        digits = getOption("out2text_coef_digits", 3),
                        format = getOption("out2text_coef_format", "f"),
                        sep = getOption("out2text_confint_sep", ", "),
                        brackets = getOption("out2text_confint_brackets", c("[", "]")),
                        standardized = FALSE,
                        group = 1L,
                        ...) {
    param_str <- lavaan::lavParseModelString(
                    model.syntax = param,
                    as.data.frame. = TRUE,
                    warn = FALSE)
    if (standardized) {
        ct <- lavaan::standardizedSolution(x,
                                           se = TRUE,
                                           zstat = FALSE,
                                           pvalue = FALSE,
                                           ci = TRUE,
                                           ...)
        ct$est <- ct$est.std
      } else {
        ct <- lavaan::parameterEstimates(x, ...)
      }
    if (lavaan::lavInspect(x, "ngroups") == 1) {
        ct$group <- 1
      }
    out0 <- ct[(ct$lhs == param_str$lhs) &
                (ct$op == param_str$op) &
                (ct$rhs == param_str$rhs) &
                (ct$group == group), c("ci.lower", "ci.upper")]
    out0 <- unname(unlist(out0))
    out1 <- formatC(out0,
                    digits = digits,
                    format = format)
    out <- paste0(brackets[1],
                  paste0(out1, collapse = sep),
                  brackets[2])
    out
  }

R/text_rsq_p.R
#' @title The P-Value of the F Test of R-square as Text
#'
#' @description Extracts the p-value of the
#' F statistics of
#' the R-square from an object
#' and returns it as a text string.
#'
#' @details Details
#'   (Include subjects for verbs.)
#'   (Use 3rd person forms for verbs.)
#'
#' @return A one-element character vector.
#'
#' @param x A supported object with a F statistic the results.
#'
#' @param digits The number of decimal places to keep. To be
#' passed to [formatC()].
#'
#' @param format The format string used in [formatC()]. Default
#' is `"f"`.
#'
#' @param italics Logical. Whether the statistics should be
#' italicized (using markdown). Default is `FALSE`.
#'
#' @param ...  Optional arguments. Ignored for now.
#'
#'
#' @author Shu Fai Cheung <https://orcid.org/0000-0002-9871-9448>
#'
#'
#' @examples
#'
#' set.seed(894252)
#' n <- 10
#' x <- rnorm(n)
#' y <- .6 * x + rnorm(n, 0, .8)
#' out <- lm(y ~ x)
#' text_rsq_p(out)
#'
#' @export

text_rsq_p <- function(x, ...) UseMethod("text_rsq_p")

#' @export
#' @rdname text_rsq_p

text_rsq_p.default <- function(x,
                             digits = getOption("out2text_test_p_digits", 3),
                             format = getOption("out2text_test_p_format", "f"),
                             italics = FALSE,
                             ...) {
    NULL
  }


#' @export
#' @rdname text_rsq_p

text_rsq_p.lm <- function(x,
                        digits = getOption("out2text_test_p_digits", 3),
                        format = getOption("out2text_test_p_format", "f"),
                        italics = FALSE,
                        ...) {
    fstat <- summary(x)$fstatistic
    if (is.null(fstat)) {
        stop("The summary of the object does not have the F statistic.")
      }
    fstatp <- stats::pf(q = fstat["value"],
                 df1 = fstat["numdf"],
                 df2 = fstat["dendf"],
                 lower.tail = FALSE)
    out <- format_pval(fstatp)
    names(out) <- NULL
    out
  }
